File: Cpp/LeetCode/715.cpp
#include "usual.h"

class RangeModule {
    map<int, int> ends;

public:
    RangeModule() {

    }
    
    void addRange(int left, int right) {
        --right;
        cout << "add(" << left << ", " << right << ")" << endl;
        auto p = ends.lower_bound(left - 1);
        while (p != ends.end()) {
            if (p->second > right + 1)
                break;
            left = min(left, p->second);
            right = max(right, p->first);
            auto tmp = p;
            ++p;
            ends.erase(tmp);
        }
        ends[right] = left;
        for (const auto& [r, l] : ends) {
            cout << "[" << l << ", " << r << "] ";
        }
        cout << endl;
    }
    
    bool queryRange(int left, int right) {
        --right;
        auto p = ends.lower_bound(right);
        cout << "query(" << left << ", " << right << "): " << (p != ends.end() && p->second <= left) << endl;
        return p != ends.end() && p->second <= left;
    }
    
    void removeRange(int left, int right) {
        --right;
        vector<pair<int, int>> add;
        auto p = ends.lower_bound(left);
        while (p != ends.end()) {
            if (p->second > right)
                break;
            if (p->second < left) {
                add.emplace_back(p->second, left - 1);
            }
            if (p->first > right) {
                add.emplace_back(right + 1, p->first);
            }
            auto tmp = p;
            ++p;
            ends.erase(tmp);
        }
        for (const auto& [l, r] : add) {
            ends[r] = l;
        }
        cout << "remove(" << left << ", " << right << ")" << endl;
        for (const auto& [r, l] : ends) {
            cout << "[" << l << ", " << r << "] ";
        }
        cout << endl;
    }
};

/**
 * Your RangeModule object will be instantiated and called as such:
 * RangeModule* obj = new RangeModule();
 * obj->addRange(left,right);
 * bool param_2 = obj->queryRange(left,right);
 * obj->removeRange(left,right);
 */


int main(int argc, char const *argv[])
{
    RangeModule* obj = new RangeModule();
    obj->addRange(14,100);
    obj->removeRange(1,8);
    obj->queryRange(77,80);
    obj->queryRange(8,43);
    obj->queryRange(4,13);
    obj->removeRange(3,9);
    obj->removeRange(45,49);
    obj->removeRange(41,90);
    obj->addRange(58,79);
    obj->addRange(4,83);
    obj->addRange(34,39);
    obj->removeRange(84,100);
    obj->addRange(8,9);
    obj->queryRange(32,56);
    obj->addRange(35,46);
    obj->addRange(9,100);
    obj->queryRange(85,99);
    obj->queryRange(23,33);
    obj->addRange(10,31);
    obj->removeRange(15,45);
    obj->removeRange(52,70);
    obj->removeRange(26,42);
    obj->queryRange(30,70);
    obj->queryRange(60,69);
    obj->addRange(10,94);
    obj->addRange(2,89);
    obj->queryRange(26,39);
    obj->addRange(46,93);
    obj->addRange(30,83);
    obj->removeRange(42,48);
    obj->addRange(47,74);
    obj->addRange(39,45);
    obj->queryRange(14,64);
    obj->removeRange(3,97);
    obj->queryRange(16,34);
    obj->removeRange(28,100);
    obj->addRange(19,37);
    obj->addRange(27,91);
    obj->queryRange(55,62);
    obj->removeRange(64,65);
    obj->removeRange(2,48);
    obj->addRange(55,78);
    obj->queryRange(21,89);
    obj->queryRange(31,76);
    obj->removeRange(13,32);
    obj->removeRange(2,84);
    obj->removeRange(21,88);
    obj->queryRange(12,31);
    obj->addRange(89,97);
    obj->removeRange(56,72);
    obj->removeRange(16,75);
    obj->queryRange(18,90);
    obj->removeRange(46,60);
    obj->removeRange(20,62);
    obj->queryRange(28,77);
    obj->addRange(5,78);
    obj->addRange(58,61);
    obj->removeRange(38,70);
    obj->queryRange(24,73);
    obj->queryRange(72,96);
    obj->addRange(5,24);
    obj->removeRange(43,49);
    obj->removeRange(2,20);
    obj->addRange(4,69);
    obj->addRange(18,98);
    obj->addRange(26,42);
    obj->addRange(14,18);
    obj->queryRange(46,58);
    obj->removeRange(16,90);
    obj->addRange(32,47);
    obj->addRange(19,36);
    obj->addRange(26,78);
    obj->queryRange(7,58);
    obj->addRange(42,54);
    obj->removeRange(42,83);
    obj->queryRange(3,83);
    obj->removeRange(54,82);
    obj->removeRange(71,91);
    obj->removeRange(22,37);
    obj->queryRange(38,94);
    obj->queryRange(20,44);
    obj->queryRange(37,89);
    obj->queryRange(15,54);
    obj->queryRange(1,64);
    obj->removeRange(63,65);
    obj->queryRange(55,58);
    obj->removeRange(23,44);
    obj->queryRange(25,87);
    obj->addRange(38,85);
    obj->queryRange(27,71);
    return 0;
}

File: Cpp/LeetCode/715.py
from typing import List


def generate_cpp_test_code(methods: List, args: List[List]):
    s = methods[0] + " *obj = new " + methods[0] + "(" + ",".join(args[0]) + ");\n"
    for method, arg in zip(methods, args):
        s += "obj->" + method + "(" + ",".join([str(a) for a in arg]) + ");\n"
    return s

if __name__ == '__main__':
    methods = ["RangeModule","addRange","removeRange","queryRange","queryRange","queryRange","removeRange","removeRange","removeRange","addRange","addRange","addRange","removeRange","addRange","queryRange","addRange","addRange","queryRange","queryRange","addRange","removeRange","removeRange","removeRange","queryRange","queryRange","addRange","addRange","queryRange","addRange","addRange","removeRange","addRange","addRange","queryRange","removeRange","queryRange","removeRange","addRange","addRange","queryRange","removeRange","removeRange","addRange","queryRange","queryRange","removeRange","removeRange","removeRange","queryRange","addRange","removeRange","removeRange","queryRange","removeRange","removeRange","queryRange","addRange","addRange","removeRange","queryRange","queryRange","addRange","removeRange","removeRange","addRange","addRange","addRange","addRange","queryRange","removeRange","addRange","addRange","addRange","queryRange","addRange","removeRange","queryRange","removeRange","removeRange","removeRange","queryRange","queryRange","queryRange","queryRange","queryRange","removeRange","queryRange","removeRange","queryRange","addRange","queryRange"]
    args = [[],[14,100],[1,8],[77,80],[8,43],[4,13],[3,9],[45,49],[41,90],[58,79],[4,83],[34,39],[84,100],[8,9],[32,56],[35,46],[9,100],[85,99],[23,33],[10,31],[15,45],[52,70],[26,42],[30,70],[60,69],[10,94],[2,89],[26,39],[46,93],[30,83],[42,48],[47,74],[39,45],[14,64],[3,97],[16,34],[28,100],[19,37],[27,91],[55,62],[64,65],[2,48],[55,78],[21,89],[31,76],[13,32],[2,84],[21,88],[12,31],[89,97],[56,72],[16,75],[18,90],[46,60],[20,62],[28,77],[5,78],[58,61],[38,70],[24,73],[72,96],[5,24],[43,49],[2,20],[4,69],[18,98],[26,42],[14,18],[46,58],[16,90],[32,47],[19,36],[26,78],[7,58],[42,54],[42,83],[3,83],[54,82],[71,91],[22,37],[38,94],[20,44],[37,89],[15,54],[1,64],[63,65],[55,58],[23,44],[25,87],[38,85],[27,71]]
    print(generate_cpp_test_code(methods, args))
